File: mycollab-mobile/src/main/scala/com/esofthead/mycollab/mobile/module/project/view/ProjectModuleController.scala
package com.esofthead.mycollab.mobile.module.project.view

import com.esofthead.mycollab.common.i18n.OptionI18nEnum.StatusI18nEnum
import com.esofthead.mycollab.core.MyCollabException
import com.esofthead.mycollab.core.arguments.{NumberSearchField, SearchField, SetSearchField, StringSearchField}
import com.esofthead.mycollab.core.utils.BeanUtility
import com.esofthead.mycollab.eventmanager.ApplicationEventListener
import com.esofthead.mycollab.mobile.module.project.events._
import com.esofthead.mycollab.mobile.module.project.view.bug.BugPresenter
import com.esofthead.mycollab.mobile.module.project.view.message.MessagePresenter
import com.esofthead.mycollab.mobile.module.project.view.milestone.MilestonePresenter
import com.esofthead.mycollab.mobile.module.project.view.parameters.ProjectScreenData.{Add, ViewActivities}
import com.esofthead.mycollab.mobile.module.project.view.parameters._
import com.esofthead.mycollab.mobile.module.project.view.settings.ProjectUserPresenter
import com.esofthead.mycollab.mobile.module.project.view.task.TaskPresenter
import com.esofthead.mycollab.module.project.domain.criteria._
import com.esofthead.mycollab.module.project.domain.{SimpleMilestone, SimpleProject, SimpleProjectMember, SimpleTask}
import com.esofthead.mycollab.module.project.{CurrentProjectVariables, ProjectMemberStatusConstants}
import com.esofthead.mycollab.module.tracker.domain.SimpleBug
import com.esofthead.mycollab.module.tracker.domain.criteria.BugSearchCriteria
import com.esofthead.mycollab.vaadin.AppContext
import com.esofthead.mycollab.vaadin.mvp.{AbstractController, PageActionChain, PresenterResolver, ScreenData}
import com.google.common.eventbus.Subscribe
import com.vaadin.addon.touchkit.ui.NavigationManager

/**
  * @author MyCollab Ltd
  * @since 5.2.5
  */
class ProjectModuleController(val navManager: NavigationManager) extends AbstractController {
  bindProjectEvents()
  bindBugEvents()
  bindMessageEvents()
  bindMilestoneEvents()
  bindTaskEvents()
  bindMemberEvents()

  private def bindProjectEvents() {
    this.register(new ApplicationEventListener[ProjectEvent.GotoAdd]() {
      @Subscribe def handle(event: ProjectEvent.GotoAdd): Unit = {
        val presenter = PresenterResolver.getPresenter(classOf[ProjectAddPresenter])
        presenter.go(navManager, new Add(new SimpleProject))
      }
    })
    this.register(new ApplicationEventListener[ProjectEvent.GotoProjectList]() {
      @Subscribe def handle(event: ProjectEvent.GotoProjectList) {
        val presenter = PresenterResolver.getPresenter(classOf[ProjectListPresenter])
        val criteria = new ProjectSearchCriteria
        criteria.setInvolvedMember(new StringSearchField(AppContext.getUsername))
        criteria.setProjectStatuses(new SetSearchField[String](StatusI18nEnum.Open.name))
        presenter.go(navManager, new ScreenData.Search[ProjectSearchCriteria](criteria))
      }
    })
    this.register(new ApplicationEventListener[ProjectEvent.GotoMyProject]() {
      @Subscribe def handle(event: ProjectEvent.GotoMyProject) {
        val presenter = PresenterResolver.getPresenter(classOf[ProjectViewPresenter])
        presenter.handleChain(navManager, event.getData.asInstanceOf[PageActionChain])
      }
    })
    this.register(new ApplicationEventListener[ProjectEvent.AllActivities]() {
      @Subscribe def handle(event: ProjectEvent.AllActivities) {
        val presenter = PresenterResolver.getPresenter(classOf[AllActivityStreamPresenter])
        presenter.go(navManager, event.getData.asInstanceOf[ProjectScreenData.AllActivities])
      }
    })
    this.register(new ApplicationEventListener[ProjectEvent.MyProjectActivities]() {
      @Subscribe def handle(event: ProjectEvent.MyProjectActivities) {
        val presenter: ProjectActivityStreamPresenter = PresenterResolver.getPresenter(classOf[ProjectActivityStreamPresenter])
        presenter.go(navManager, new ViewActivities(event.getData.asInstanceOf[Integer]))
      }
    })
  }

  private def bindBugEvents() {
    this.register(new ApplicationEventListener[BugEvent.GotoList]() {
      @Subscribe def handle(event: BugEvent.GotoList) {
        val params: Any = event.getData
        val presenter = PresenterResolver.getPresenter(classOf[BugPresenter])
        if (params == null) {
          val criteria = new BugSearchCriteria
          criteria.setProjectId(new NumberSearchField(SearchField.AND, CurrentProjectVariables.getProjectId))
          presenter.go(navManager, new BugScreenData.Search(criteria))
        }
        else if (params.isInstanceOf[BugScreenData.Search]) {
          presenter.go(navManager, params.asInstanceOf[BugScreenData.Search])
        }
        else {
          throw new MyCollabException("Invalid search parameter: " + BeanUtility.printBeanObj(params))
        }
      }
    })
    this.register(new ApplicationEventListener[BugEvent.GotoRead]() {
      @Subscribe def handle(event: BugEvent.GotoRead) {
        val data = new BugScreenData.Read(event.getData.asInstanceOf[Integer])
        val presenter = PresenterResolver.getPresenter(classOf[BugPresenter])
        presenter.go(navManager, data)
      }
    })
    this.register(new ApplicationEventListener[BugEvent.GotoAdd]() {
      @Subscribe def handle(event: BugEvent.GotoAdd) {
        val data = new BugScreenData.Add(new SimpleBug)
        val presenter = PresenterResolver.getPresenter(classOf[BugPresenter])
        presenter.go(navManager, data)
      }
    })
    this.register(new ApplicationEventListener[BugEvent.GotoEdit]() {
      @Subscribe def handle(event: BugEvent.GotoEdit) {
        val data: BugScreenData.Edit = new BugScreenData.Edit(event.getData.asInstanceOf[SimpleBug])
        val presenter: BugPresenter = PresenterResolver.getPresenter(classOf[BugPresenter])
        presenter.go(navManager, data)
      }
    })
  }

  private def bindMessageEvents() {
    this.register(new ApplicationEventListener[MessageEvent.GotoAdd]() {
      @Subscribe def handle(event: MessageEvent.GotoAdd) {
        val data = new MessageScreenData.Add
        val presenter = PresenterResolver.getPresenter(classOf[MessagePresenter])
        presenter.go(navManager, data)
      }
    })
    this.register(new ApplicationEventListener[MessageEvent.GotoList]() {
      @Subscribe def handle(event: MessageEvent.GotoList) {
        val searchCriteria = new MessageSearchCriteria
        searchCriteria.setProjectids(new SetSearchField[Integer](CurrentProjectVariables.getProjectId))
        val data = new MessageScreenData.Search(searchCriteria)
        val presenter = PresenterResolver.getPresenter(classOf[MessagePresenter])
        presenter.go(navManager, data)
      }
    })
    this.register(new ApplicationEventListener[MessageEvent.GotoRead]() {
      @Subscribe def handle(event: MessageEvent.GotoRead) {
        val data = new MessageScreenData.Read(event.getData.asInstanceOf[Integer])
        val presenter = PresenterResolver.getPresenter(classOf[MessagePresenter])
        presenter.go(navManager, data)
      }
    })
  }

  private def bindMilestoneEvents() {
    this.register(new ApplicationEventListener[MilestoneEvent.GotoList]() {
      @Subscribe def handle(event: MilestoneEvent.GotoList) {
        val params: Any = event.getData
        val presenter: MilestonePresenter = PresenterResolver.getPresenter(classOf[MilestonePresenter])
        if (params == null) {
          val criteria = new MilestoneSearchCriteria
          criteria.setProjectId(new NumberSearchField(SearchField.AND, CurrentProjectVariables.getProjectId))
          presenter.go(navManager, new MilestoneScreenData.Search(criteria))
        }
        else if (params.isInstanceOf[MilestoneScreenData.Search]) {
          presenter.go(navManager, params.asInstanceOf[MilestoneScreenData.Search])
        }
        else {
          throw new MyCollabException("Invalid search parameter: " + BeanUtility.printBeanObj(params))
        }
      }
    })
    this.register(new ApplicationEventListener[MilestoneEvent.GotoRead]() {
      @Subscribe def handle(event: MilestoneEvent.GotoRead) {
        val data = new MilestoneScreenData.Read(event.getData.asInstanceOf[Integer])
        val presenter = PresenterResolver.getPresenter(classOf[MilestonePresenter])
        presenter.go(navManager, data)
      }
    })
    this.register(new ApplicationEventListener[MilestoneEvent.GotoAdd]() {
      @Subscribe def handle(event: MilestoneEvent.GotoAdd) {
        val data = new MilestoneScreenData.Add(new SimpleMilestone)
        val presenter = PresenterResolver.getPresenter(classOf[MilestonePresenter])
        presenter.go(navManager, data)
      }
    })
    this.register(new ApplicationEventListener[MilestoneEvent.GotoEdit]() {
      @Subscribe def handle(event: MilestoneEvent.GotoEdit) {
        val data = new MilestoneScreenData.Edit(event.getData.asInstanceOf[SimpleMilestone])
        val presenter = PresenterResolver.getPresenter(classOf[MilestonePresenter])
        presenter.go(navManager, data)
      }
    })
  }

  private def bindTaskEvents() {
    this.register(new ApplicationEventListener[TaskEvent.GotoList]() {
      @Subscribe def handle(event: TaskEvent.GotoList) {
        val criteria = new TaskSearchCriteria
        criteria.setProjectid(new NumberSearchField(SearchField.AND, CurrentProjectVariables.getProjectId))
        val data = new TaskScreenData.Search(criteria)
        val presenter = PresenterResolver.getPresenter(classOf[TaskPresenter])
        presenter.go(navManager, data)
      }
    })
    this.register(new ApplicationEventListener[TaskEvent.GotoRead]() {
      @Subscribe def handle(event: TaskEvent.GotoRead) {
        val data = new TaskScreenData.Read(event.getData.asInstanceOf[Integer])
        val presenter = PresenterResolver.getPresenter(classOf[TaskPresenter])
        presenter.go(navManager, data)
      }
    })
    this.register(new ApplicationEventListener[TaskEvent.GotoEdit]() {
      @Subscribe def handle(event: TaskEvent.GotoEdit) {
        val data = new TaskScreenData.Edit(event.getData.asInstanceOf[SimpleTask])
        val presenter = PresenterResolver.getPresenter(classOf[TaskPresenter])
        presenter.go(navManager, data)
      }
    })
    this.register(new ApplicationEventListener[TaskEvent.GotoAdd]() {
      @Subscribe def handle(event: TaskEvent.GotoAdd) {
        val data = new TaskScreenData.Add(new SimpleTask)
        val presenter = PresenterResolver.getPresenter(classOf[TaskPresenter])
        presenter.go(navManager, data)
      }
    })
  }

  private def bindMemberEvents() {
    this.register(new ApplicationEventListener[ProjectMemberEvent.GotoList]() {
      @Subscribe def handle(event: ProjectMemberEvent.GotoList) {
        val criteria = new ProjectMemberSearchCriteria
        criteria.setProjectId(new NumberSearchField(CurrentProjectVariables.getProjectId))
        criteria.setSaccountid(new NumberSearchField(AppContext.getAccountId))
        criteria.setStatus(new StringSearchField(ProjectMemberStatusConstants.ACTIVE))
        val presenter = PresenterResolver.getPresenter(classOf[ProjectUserPresenter])
        presenter.go(navManager, new ProjectMemberScreenData.Search(criteria))
      }
    })
    this.register(new ApplicationEventListener[ProjectMemberEvent.GotoRead]() {
      @Subscribe def handle(event: ProjectMemberEvent.GotoRead) {
        val data = new ProjectMemberScreenData.Read(event.getData)
        val presenter = PresenterResolver.getPresenter(classOf[ProjectUserPresenter])
        presenter.go(navManager, data)
      }
    })
    this.register(new ApplicationEventListener[ProjectMemberEvent.GotoEdit]() {
      @Subscribe def handle(event: ProjectMemberEvent.GotoEdit) {
        val data = new ProjectMemberScreenData.Edit(event.getData.asInstanceOf[SimpleProjectMember])
        val presenter = PresenterResolver.getPresenter(classOf[ProjectUserPresenter])
        presenter.go(navManager, data)
      }
    })
    this.register(new ApplicationEventListener[ProjectMemberEvent.GotoInviteMembers]() {
      @Subscribe def handle(event: ProjectMemberEvent.GotoInviteMembers) {
        val data = new ProjectMemberScreenData.InviteProjectMembers
        val presenter = PresenterResolver.getPresenter(classOf[ProjectUserPresenter])
        presenter.go(navManager, data)
      }
    })
  }
}

File: mycollab-mobile/src/main/java/com/mycollab/mobile/module/project/view/ProjectModuleController.kt
package com.mycollab.mobile.module.project.view

import com.google.common.eventbus.Subscribe
import com.mycollab.common.ModuleNameConstants
import com.mycollab.common.domain.criteria.ActivityStreamSearchCriteria
import com.mycollab.common.i18n.OptionI18nEnum.StatusI18nEnum
import com.mycollab.core.MyCollabException
import com.mycollab.db.arguments.NumberSearchField
import com.mycollab.db.arguments.SearchField
import com.mycollab.db.arguments.SetSearchField
import com.mycollab.db.arguments.StringSearchField
import com.mycollab.eventmanager.ApplicationEventListener
import com.mycollab.mobile.module.project.event.*
import com.mycollab.mobile.module.project.event.TicketEvent
import com.mycollab.mobile.module.project.view.message.MessagePresenter
import com.mycollab.mobile.module.project.view.milestone.MilestonePresenter
import com.mycollab.mobile.module.project.view.parameters.*
import com.mycollab.mobile.module.project.view.settings.ProjectUserPresenter
import com.mycollab.mobile.module.project.view.ticket.TicketPresenter
import com.mycollab.mobile.mvp.view.PresenterOptionUtil
import com.mycollab.module.project.CurrentProjectVariables
import com.mycollab.module.project.ProjectMemberStatusConstants
import com.mycollab.module.project.domain.*
import com.mycollab.module.project.domain.criteria.*
import com.mycollab.module.project.service.ProjectService
import com.mycollab.module.tracker.domain.SimpleBug
import com.mycollab.spring.AppContextUtil
import com.mycollab.vaadin.AppUI
import com.mycollab.vaadin.UserUIContext
import com.mycollab.vaadin.mvp.AbstractController
import com.mycollab.vaadin.mvp.PageActionChain
import com.mycollab.vaadin.mvp.PresenterResolver
import com.mycollab.vaadin.mvp.ScreenData
import com.vaadin.addon.touchkit.ui.NavigationManager

/**
 * @author MyCollab Ltd
 * @since 6.0.0
 */
class ProjectModuleController(val navManager: NavigationManager) : AbstractController() {
    init {
        bindProjectEvents()
        bindTicketEvents()
        bindBugEvents()
        bindMessageEvents()
        bindMilestoneEvents()
        bindTaskEvents()
        bindRiskEvents()
        bindMemberEvents()
    }

    private fun bindProjectEvents() {
        this.register(object : ApplicationEventListener<ProjectEvent.GotoAdd> {
            @Subscribe
            override fun handle(event: ProjectEvent.GotoAdd) {
                val presenter = PresenterOptionUtil.getPresenter(IProjectAddPresenter::class.java)
                presenter.go(navManager, ProjectScreenData.Add(SimpleProject()))
            }
        })
        this.register(object : ApplicationEventListener<ProjectEvent.GotoProjectList> {
            @Subscribe
            override fun handle(event: ProjectEvent.GotoProjectList) {
                val presenter = PresenterResolver.getPresenter(UserProjectListPresenter::class.java)
                val criteria = ProjectSearchCriteria()
                criteria.involvedMember = StringSearchField.and(UserUIContext.getUsername())
                criteria.projectStatuses = SetSearchField<String>(StatusI18nEnum.Open.name)
                presenter.go(navManager, ScreenData.Search<ProjectSearchCriteria>(criteria))
            }
        })
        this.register(object : ApplicationEventListener<ProjectEvent.GotoMyProject> {
            @Subscribe
            override fun handle(event: ProjectEvent.GotoMyProject) {
                val presenter = PresenterResolver.getPresenter(ProjectViewPresenter::class.java)
                presenter.handleChain(navManager, event.data as PageActionChain)
            }
        })
        this.register(object : ApplicationEventListener<ProjectEvent.GotoAllActivitiesView> {
            @Subscribe
            override fun handle(event: ProjectEvent.GotoAllActivitiesView) {
                val presenter = PresenterResolver.getPresenter(AllActivitiesStreamPresenter::class.java)
                val prjService = AppContextUtil.getSpringBean(ProjectService::class.java)
                val prjKeys = prjService.getProjectKeysUserInvolved(UserUIContext.getUsername(), AppUI.accountId)
                val searchCriteria = ActivityStreamSearchCriteria()
                searchCriteria.moduleSet = SetSearchField(ModuleNameConstants.PRJ)
                searchCriteria.saccountid = NumberSearchField(AppUI.accountId)
                searchCriteria.extraTypeIds = SetSearchField(prjKeys)
                presenter.go(navManager, ProjectScreenData.AllActivities(searchCriteria))
            }
        })
        this.register(object : ApplicationEventListener<ProjectEvent.MyProjectActivities> {
            @Subscribe
            override fun handle(event: ProjectEvent.MyProjectActivities) {
                val presenter = PresenterResolver.getPresenter(ProjectActivityStreamPresenter::class.java)
                val searchCriteria = ActivityStreamSearchCriteria()
                searchCriteria.moduleSet = SetSearchField(ModuleNameConstants.PRJ)
                searchCriteria.saccountid = NumberSearchField(AppUI.accountId)
                searchCriteria.extraTypeIds = SetSearchField(event.data as Int)
                presenter.go(navManager, ProjectScreenData.ProjectActivities(searchCriteria))
            }
        })
    }

    private fun bindTicketEvents() {
        this.register(object : ApplicationEventListener<TicketEvent.GotoDashboard> {
            @Subscribe
            override fun handle(event: TicketEvent.GotoDashboard) {
                val searchCriteria = ProjectTicketSearchCriteria()
                searchCriteria.projectIds = SetSearchField<Int>(CurrentProjectVariables.projectId)
                searchCriteria.types = CurrentProjectVariables.restrictedTicketTypes
                searchCriteria.isOpenned = SearchField()
                val data = TicketScreenData.GotoDashboard(searchCriteria)
                val presenter = PresenterResolver.getPresenter(TicketPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
    }

    private fun bindBugEvents() {
        this.register(object : ApplicationEventListener<BugEvent.GotoRead> {
            @Subscribe
            override fun handle(event: BugEvent.GotoRead) {
                val data = BugScreenData.Read(event.data as Int)
                val presenter = PresenterResolver.getPresenter(TicketPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
        this.register(object : ApplicationEventListener<BugEvent.GotoAdd> {
            @Subscribe
            override fun handle(event: BugEvent.GotoAdd) {
                val data = BugScreenData.Add(SimpleBug())
                val presenter = PresenterResolver.getPresenter(TicketPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
        this.register(object : ApplicationEventListener<BugEvent.GotoEdit> {
            @Subscribe
            override fun handle(event: BugEvent.GotoEdit) {
                val data = BugScreenData.Edit(event.data as SimpleBug)
                val presenter = PresenterResolver.getPresenter(TicketPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
    }

    private fun bindMessageEvents() {
        this.register(object : ApplicationEventListener<MessageEvent.GotoAdd> {
            @Subscribe
            override fun handle(event: MessageEvent.GotoAdd) {
                val presenter = PresenterResolver.getPresenter(MessagePresenter::class.java)
                presenter.go(navManager, MessageScreenData.Add())
            }
        })
        this.register(object : ApplicationEventListener<MessageEvent.GotoList> {
            @Subscribe
            override fun handle(event: MessageEvent.GotoList) {
                val searchCriteria = MessageSearchCriteria()
                searchCriteria.projectids = SetSearchField<Int>(CurrentProjectVariables.projectId)
                val data = MessageScreenData.Search(searchCriteria)
                val presenter = PresenterResolver.getPresenter(MessagePresenter::class.java)
                presenter.go(navManager, data)
            }
        })
        this.register(object : ApplicationEventListener<MessageEvent.GotoRead> {
            @Subscribe
            override fun handle(event: MessageEvent.GotoRead) {
                val presenter = PresenterResolver.getPresenter(MessagePresenter::class.java)
                presenter.go(navManager, MessageScreenData.Read(event.data as Int))
            }
        })
    }

    private fun bindMilestoneEvents() {
        this.register(object : ApplicationEventListener<MilestoneEvent.GotoList> {
            @Subscribe
            override fun handle(event: MilestoneEvent.GotoList) {
                val params = event.data
                val presenter = PresenterResolver.getPresenter(MilestonePresenter::class.java)
                if (params == null) {
                    val criteria = MilestoneSearchCriteria()
                    criteria.projectIds = SetSearchField<Int>(CurrentProjectVariables.projectId)
                    presenter.go(navManager, MilestoneScreenData.Search(criteria))
                } else if (params is MilestoneScreenData.Search) {
                    presenter.go(navManager, params)
                } else {
                    throw MyCollabException("Invalid search parameter $params")
                }
            }
        })
        this.register(object : ApplicationEventListener<MilestoneEvent.GotoRead> {
            @Subscribe
            override fun handle(event: MilestoneEvent.GotoRead) {
                val data = MilestoneScreenData.Read(event.data as Int)
                val presenter = PresenterResolver.getPresenter(MilestonePresenter::class.java)
                presenter.go(navManager, data)
            }
        })
        this.register(object : ApplicationEventListener<MilestoneEvent.GotoAdd> {
            @Subscribe
            override fun handle(event: MilestoneEvent.GotoAdd) {
                val data = MilestoneScreenData.Add(SimpleMilestone())
                val presenter = PresenterResolver.getPresenter(MilestonePresenter::class.java)
                presenter.go(navManager, data)
            }
        })
        this.register(object : ApplicationEventListener<MilestoneEvent.GotoEdit> {
            @Subscribe
            override fun handle(event: MilestoneEvent.GotoEdit) {
                val data = MilestoneScreenData.Edit(event.data as SimpleMilestone)
                val presenter = PresenterResolver.getPresenter(MilestonePresenter::class.java)
                presenter.go(navManager, data)
            }
        })
    }

    private fun bindTaskEvents() {
        this.register(object : ApplicationEventListener<TaskEvent.GotoRead> {
            @Subscribe
            override fun handle(event: TaskEvent.GotoRead) {
                val data = TaskScreenData.Read(event.data as Int)
                val presenter = PresenterResolver.getPresenter(TicketPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
        this.register(object : ApplicationEventListener<TaskEvent.GotoEdit> {
            @Subscribe
            override fun handle(event: TaskEvent.GotoEdit) {
                val data = TaskScreenData.Edit(event.data as SimpleTask)
                val presenter = PresenterResolver.getPresenter(TicketPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
        this.register(object : ApplicationEventListener<TaskEvent.GotoAdd> {
            @Subscribe
            override fun handle(event: TaskEvent.GotoAdd) {
                val data = TaskScreenData.Add(SimpleTask())
                val presenter = PresenterResolver.getPresenter(TicketPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
    }

    private fun bindRiskEvents() {
        this.register(object : ApplicationEventListener<RiskEvent.GotoRead> {
            @Subscribe
            override fun handle(event: RiskEvent.GotoRead) {
                val data = RiskScreenData.Read(event.data as Int)
                val presenter = PresenterOptionUtil.getPresenter(TicketPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
        this.register(object : ApplicationEventListener<RiskEvent.GotoAdd> {
            @Subscribe
            override fun handle(event: RiskEvent.GotoAdd) {
                val data = RiskScreenData.Add(SimpleRisk())
                val presenter = PresenterOptionUtil.getPresenter(TicketPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
        this.register(object : ApplicationEventListener<RiskEvent.GotoEdit> {
            @Subscribe
            override fun handle(event: RiskEvent.GotoEdit) {
                val data = RiskScreenData.Edit(event.data as SimpleRisk)
                val presenter = PresenterOptionUtil.getPresenter(TicketPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
    }

    private fun bindMemberEvents() {
        this.register(object : ApplicationEventListener<ProjectMemberEvent.GotoList> {
            @Subscribe
            override fun handle(event: ProjectMemberEvent.GotoList) {
                val criteria = ProjectMemberSearchCriteria()
                criteria.projectId = NumberSearchField.equal(CurrentProjectVariables.projectId)
                criteria.saccountid = NumberSearchField.equal(AppUI.accountId)
                criteria.statuses = SetSearchField(ProjectMemberStatusConstants.ACTIVE, ProjectMemberStatusConstants.NOT_ACCESS_YET)
                val presenter = PresenterResolver.getPresenter(ProjectUserPresenter::class.java)
                presenter.go(navManager, ProjectMemberScreenData.Search(criteria))
            }
        })
        this.register(object : ApplicationEventListener<ProjectMemberEvent.GotoRead> {
            @Subscribe
            override fun handle(event: ProjectMemberEvent.GotoRead) {
                val data = ProjectMemberScreenData.Read(event.data)
                val presenter = PresenterResolver.getPresenter(ProjectUserPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
        this.register(object : ApplicationEventListener<ProjectMemberEvent.GotoEdit> {
            @Subscribe
            override fun handle(event: ProjectMemberEvent.GotoEdit) {
                val data = ProjectMemberScreenData.Edit(event.data as SimpleProjectMember)
                val presenter = PresenterResolver.getPresenter(ProjectUserPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
        this.register(object : ApplicationEventListener<ProjectMemberEvent.GotoInviteMembers> {
            @Subscribe
            override fun handle(event: ProjectMemberEvent.GotoInviteMembers) {
                val data = ProjectMemberScreenData.InviteProjectMembers()
                val presenter = PresenterResolver.getPresenter(ProjectUserPresenter::class.java)
                presenter.go(navManager, data)
            }
        })
    }
}